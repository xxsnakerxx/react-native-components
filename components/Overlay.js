import React from 'react';
import PropTypes from 'prop-types';

import {
  StyleSheet,
  Animated,
  View,
  ViewPropTypes,
  TouchableWithoutFeedback,
  ActivityIndicator,
  StatusBar,
} from 'react-native';

import Modal from './Modal';

export default class Overlay extends React.Component {
  static propTypes = {
    style: ViewPropTypes.style,
    modal: PropTypes.bool,
    spinner: PropTypes.bool,
    spinnerColor: PropTypes.string,
    backgroundColor: PropTypes.string,
    visible: PropTypes.bool,
    layer: PropTypes.object,
    onPress: PropTypes.func,
  }

  static defaultProps = {
    style: null,
    modal: false,
    spinner: false,
    visible: false,
    spinnerColor: 'white',
    backgroundColor: 'rgba(0, 0, 0, 0.75)',
    layer: null,
    onPress: () => {},
  }

  constructor(props) {
    super(props);

    this.state = {
      visible: false,
      opacity: new Animated.Value(0),
    };
  }

  componentDidMount() {
    const {
      visible,
    } = this.props;

    if (visible) this.show();
  }

  componentDidUpdate(prevProps) {
    const {
      visible,
    } = this.props;

    if (visible && !prevProps.visible) {
      this.show();
    }

    if (!visible && prevProps.visible) {
      this.hide();
    }
  }

  componentWillUnmount() {
    const {
      opacity,
    } = this.state;

    opacity.stopAnimation();
  }

  _renderSpinner = () => {
    const { spinnerColor } = this.props;

    return (
      <View style={styles.spinnerContainer}>
        <ActivityIndicator color={spinnerColor} size="large" />
      </View>
    );
  };

  _onPress = () => {
    const {
      onPress,
    } = this.props;

    onPress();
  }

  _renderOverlay = () => {
    const {
      style,
      layer,
      spinner,
      modal,
      backgroundColor,
      children,
    } = this.props;

    const {
      visible,
      opacity,
    } = this.state;

    return (
      <View
        pointerEvents={visible ? 'auto' : 'none'}
        removeClippedSubviews={!visible}
        style={[
          styles.overlay,
          // eslint-disable-next-line react-native/no-inline-styles
          !modal ? {
            opacity: +!!visible,
            overflow: visible ? 'visible' : 'hidden',
          } : {},
        ]}
      >
        <StatusBar
          animated
          barStyle="light-content"
          backgroundColor={backgroundColor}
        />
        <Animated.View
          style={[
            styles.overlay,
            // eslint-disable-next-line react-native/no-inline-styles
            {
              backgroundColor:
              layer && React.isValidElement(layer)
                ? 'transparent'
                : backgroundColor,
            },
            style,
            {
              opacity,

            },
          ]}
        >
          {layer && React.isValidElement(layer)
            ? React.cloneElement(layer, {
              style: styles.overlay,
            })
            : null
          }
          {spinner
            ? this._renderSpinner()
            : null
          }
        </Animated.View>
        {!spinner
          ? (
            <View style={styles.flexContainer}>
              <TouchableWithoutFeedback
                onPress={this._onPress}
                style={styles.overlay}
              >
                <View style={styles.flexContainer} />
              </TouchableWithoutFeedback>
              {children}
            </View>
          )
          : null
        }
      </View>
    );
  }

  show(cb) {
    const {
      opacity,
    } = this.state;

    this.setState({
      visible: true,
    });

    Animated.timing(opacity, {
      toValue: 1,
      duration: 300,
      useNativeDriver: true,
    }).start(() => {
      if (cb) cb();
    });
  }

  hide(cb) {
    const {
      opacity,
    } = this.state;

    Animated.timing(opacity, {
      toValue: 0,
      duration: 300,
      useNativeDriver: true,
    }).start(({ finished }) => {
      if (finished) {
        this.setState({
          visible: false,
        });
      }

      if (cb) cb();
    });
  }

  render() {
    const {
      modal,
    } = this.props;

    const {
      visible,
    } = this.state;

    return !modal
      ? this._renderOverlay()
      : (
        <Modal visible={visible}>
          {this._renderOverlay()}
        </Modal>
      );
  }
}

const styles = StyleSheet.create({
  overlay: {
    ...StyleSheet.absoluteFillObject,
  },
  spinnerContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  flexContainer: {
    flex: 1,
  },
});
