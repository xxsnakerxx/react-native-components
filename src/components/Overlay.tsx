import React from 'react';

import {
  StyleSheet,
  Animated,
  View,
  TouchableWithoutFeedback,
  ActivityIndicator,
  StatusBar,
  StyleProp,
  ViewStyle,
  GestureResponderEvent,
} from 'react-native';

import Modal from './Modal';

export interface Props {
  isVisible?: boolean;
  isModal?: boolean;
  showSpinner?: boolean;
  spinnerColor?: string;
  backgroundColor?: string;
  style?: StyleProp<ViewStyle>;
  onPress?: (e: GestureResponderEvent) => void;
  onShow?: () => void;
  onHide?: () => void;
}

interface State {
  isVisible: boolean;
  opacity: Animated.Value;
}

export default class Overlay extends React.PureComponent<Props, State> {
  static defaultProps = {
    isVisible: false,
    isModal: false,
    showSpinner: false,
    spinnerColor: 'white',
    backgroundColor: 'rgba(0, 0, 0, 0.75)',
    onPress: () => {},
    onShow: () => {},
    onHide: () => {},
  };

  constructor(props: Props) {
    super(props);

    this.state = {
      isVisible: false,
      opacity: new Animated.Value(0),
    };
  }

  componentDidMount() {
    const {isVisible} = this.props;

    if (isVisible) {
      this.show();
    }
  }

  componentDidUpdate(prevProps: Props) {
    const {isVisible} = this.props;

    if (isVisible && !prevProps.isVisible) {
      this.show();
    }

    if (!isVisible && prevProps.isVisible) {
      this.hide();
    }
  }

  componentWillUnmount() {
    const {opacity} = this.state;

    opacity.stopAnimation();
  }

  _renderSpinner = () => {
    const {spinnerColor} = this.props;

    return (
      <View testID="SpinnerContainer" style={styles.spinnerContainer}>
        <ActivityIndicator color={spinnerColor} size="large" />
      </View>
    );
  };

  _onPress = (e: GestureResponderEvent) => {
    const {onPress} = this.props;

    onPress!(e);
  };

  _renderOverlay = () => {
    const {style, showSpinner, isModal, backgroundColor, children} = this.props;

    const {isVisible, opacity} = this.state;

    return (
      <View
        testID="Overlay"
        pointerEvents={isVisible ? 'auto' : 'none'}
        style={[
          styles.overlay,
          !isModal
            ? // eslint-disable-next-line react-native/no-inline-styles
              {
                opacity: +!!isVisible,
                overflow: isVisible ? 'visible' : 'hidden',
              }
            : {},
        ]}>
        {isModal ? (
          <StatusBar
            animated
            barStyle="light-content"
            backgroundColor={backgroundColor}
          />
        ) : null}
        <Animated.View
          style={[styles.overlay, {backgroundColor}, style, {opacity}]}>
          {showSpinner ? this._renderSpinner() : null}
        </Animated.View>
        {!showSpinner ? (
          <View style={styles.flex1}>
            <TouchableWithoutFeedback
              testID="PressHandler"
              onPress={this._onPress}
              style={styles.overlay}>
              <View style={styles.flex1} />
            </TouchableWithoutFeedback>
            {children}
          </View>
        ) : null}
      </View>
    );
  };

  show() {
    const {onShow} = this.props;

    const {opacity} = this.state;

    this.setState({
      isVisible: true,
    });

    Animated.timing(opacity, {
      toValue: 1,
      duration: 300,
      useNativeDriver: true,
    }).start(() => {
      onShow!();
    });
  }

  hide() {
    const {onHide} = this.props;

    const {opacity} = this.state;

    Animated.timing(opacity, {
      toValue: 0,
      duration: 300,
      useNativeDriver: true,
    }).start(({finished}) => {
      if (finished) {
        this.setState({
          isVisible: false,
        });

        onHide!();
      }
    });
  }

  render() {
    const {isModal} = this.props;

    const {isVisible} = this.state;

    return !isModal ? (
      this._renderOverlay()
    ) : (
      <Modal isVisible={isVisible} isAnimated={false}>
        {this._renderOverlay()}
      </Modal>
    );
  }
}

const styles = StyleSheet.create({
  flex1: {
    flex: 1,
  },
  overlay: {
    ...StyleSheet.absoluteFillObject,
  },
  spinnerContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
});
