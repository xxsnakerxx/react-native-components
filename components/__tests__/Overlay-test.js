import {
  Text,
} from 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import Overlay from '../Overlay';

describe('Overlay', () => {
  const style = { width: 200 };

  it('renders correctly with defaults', () => {
    const tree = renderer.create(
      <Overlay />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly as modal', () => {
    const tree = renderer.create(
      <Overlay
        modal
      />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with layer', () => {
    const tree = renderer.create(
      <Overlay
        style={style}
        layer={(
          <Text>
            Layer
          </Text>
        )}
      />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with spinner', () => {
    const tree = renderer.create(
      <Overlay
        spinner
        spinnerColor="red"
      />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when visible', () => {
    jest.useFakeTimers();

    const tree = renderer.create(
      <Overlay
        visible
      />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
