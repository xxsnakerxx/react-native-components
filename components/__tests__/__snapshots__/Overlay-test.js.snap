// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overlay renders correctly as modal 1`] = `null`;

exports[`Overlay renders correctly when visible 1`] = `
<View
  pointerEvents="auto"
  removeClippedSubviews={false}
  style={
    Array [
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      Object {
        "opacity": 1,
        "overflow": "visible",
      },
    ]
  }
>
  <View
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.75)",
        "bottom": 0,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      accessible={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </View>
</View>
`;

exports[`Overlay renders correctly with defaults 1`] = `
<View
  pointerEvents="none"
  removeClippedSubviews={true}
  style={
    Array [
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      Object {
        "opacity": 0,
        "overflow": "hidden",
      },
    ]
  }
>
  <View
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.75)",
        "bottom": 0,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      accessible={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </View>
</View>
`;

exports[`Overlay renders correctly with layer 1`] = `
<View
  pointerEvents="none"
  removeClippedSubviews={true}
  style={
    Array [
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      Object {
        "opacity": 0,
        "overflow": "hidden",
      },
    ]
  }
>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
        "bottom": 0,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": 200,
      }
    }
  >
    <Text
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      Layer
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      accessible={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </View>
</View>
`;

exports[`Overlay renders correctly with spinner 1`] = `
<View
  pointerEvents="none"
  removeClippedSubviews={true}
  style={
    Array [
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      Object {
        "opacity": 0,
        "overflow": "hidden",
      },
    ]
  }
>
  <View
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.75)",
        "bottom": 0,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <ActivityIndicator
        animating={true}
        color="red"
        hidesWhenStopped={true}
        size="large"
      />
    </View>
  </View>
</View>
`;
